loopout.go

```diff
@@ -975,6 +983,92 @@ func (s *loopOutSwap) pushPreimage(ctx context.Context) {
 	}
 }
 
+// failOffChain updates a swap's state when it has failed due to a routing
+// failure and notifies the server of the failure.
+func (s *loopOutSwap) failOffChain(ctx context.Context, paymentType paymentType,
+	status *lndclient.PaymentStatus) {
+
+	// Set our state to failed off chain timeout.
+	s.state = loopdb.StateFailOffchainPayments
+
+	swapPayReq, err := zpay32.Decode(
+		s.LoopOutContract.SwapInvoice, s.swapConfig.lnd.ChainParams,
+	)
+	if err != nil {
+		s.log.Errorf("could not decode swap invoice: %v", err)
+		return
+	}
+
+	if swapPayReq.PaymentAddr == nil {
+		s.log.Errorf("expected payment address for invoice")
+		return
+	}
+
+	details := &outCancelDetails{
+		hash:        s.hash,
+		paymentAddr: *swapPayReq.PaymentAddr,
+		metadata: &routeCancelMetadata{
+			paymentType: paymentType,
+		},
+	}
+
+	// It's possible that we don't have any payment status reported, in
+	// which case we will not report any htlc attempts. We set our failure
+	// reason if it is available.
+	var htlcs []*lndclient.HtlcAttempt
+	if status != nil {
+		htlcs = status.Htlcs
+		details.metadata.failureReason = status.FailureReason
+	}
+
+	for _, htlc := range htlcs {
+		if htlc.Status != lnrpc.HTLCAttempt_FAILED {
+			continue
+		}
+
+		if htlc.Route == nil {
+			continue
+		}
+
+		if len(htlc.Route.Hops) == 0 {
+			continue
+		}
+
+		if htlc.Failure == nil {
+			continue
+		}
+
+		failureIdx := htlc.Failure.FailureSourceIndex
+		hops := uint32(len(htlc.Route.Hops))
+
+		// We really don't expect a failure index that is greater than
+		// our number of hops, but we don't want to underflow so we
+		// check and log a warning if this happens.
+		if failureIdx > hops {
```

Comment:
Greater or equal? Since `FailureSourceIndex` sounds zero-based while `hops` is a length? Or is this a special case for HTLC routes where index 0 is always the own node (which wouldn't be in the list of hops)?

Reply:
Yeah indeed, nice catch!

Reply:
Okay, cool. Good to know. Maybe we could even extend the comment somewhat with what you wrote above?

Reply:
Updated comment!